backend-config/namespace_config.go

```diff
@@ -109,17 +121,32 @@ func (nc *namespaceConfig) getFromAPI(ctx context.Context) (map[string]ConfigT,
 	err = jsonfast.Unmarshal(respBody, &workspacesConfig)
 	if err != nil {
 		nc.logger.Errorf("Error while parsing request: %v", err)
-		return config, err
+		return configOnError, err
+	}
+
+	var deletedWorkspaces []string
+
+	if !nc.useIncrementalConfigUpdates {
+		nc.workspacesConfig = make(map[string]ConfigT)
 	}
 
 	for workspaceID, wc := range workspacesConfig {
+		if len(wc.Sources) == 0 {
+			deletedWorkspaces = append(deletedWorkspaces, workspaceID)
+		}
 		// always set connection flags to true for hosted and multi-tenant warehouse service
 		wc.ConnectionFlags.URL = nc.cpRouterURL
 		wc.ConnectionFlags.Services = map[string]bool{"warehouse": true}
-		workspacesConfig[workspaceID] = wc
+		nc.workspacesConfig[workspaceID] = wc
+		if wc.UpdatedAt.After(nc.lastUpdatedAt) {
+			nc.lastUpdatedAt = wc.UpdatedAt
+		}
+	}
+	for _, workspaceID := range deletedWorkspaces {
+		delete(nc.workspacesConfig, workspaceID)
 	}
```

Comment:
Are you sure this is going to work? I was expecting something like:

```go
for workspaceID := range nc.workspacesConfig {
    if _, ok := workspacesConfig[workspaceID]; !ok {
        // delete!
    }
}
```

Reply:
Yeah, I checked this (both before making the change) , and through the unit tests. Basically, if there is a key, value pair in the json with the value as **null** it gets parsed and the corresponding struct gets default initialized. 

See this snippet on go playground https://go.dev/play/p/y5OawhHI02u

Reply:
Hold on, I was referring to the third case.

1. the ID is there and the value is a valid object
2. the ID is there and the value is NULL (which is what you are referring to I guess)
3. the ID is NOT there

In regards to 2 it might help to have a `map[string]*ConfigT` instead of `map[string]ConfigT`. Have you tried that?

Reply:
For clarity this is what I mean: https://github.com/rudderlabs/rudder-server/compare/feat.useUpdateAfterForEfficientPolling...feat.useUpdateAfterForEfficientPolling-2?expand=1

If it makes sense please update the tests accordingly.

Reply:
Done